### ISSUE TITLE
clues to find a spec for RTL8168D

### ISSUE THREAD

**Issue by woreno** [Microsoft/NetAdapter-Cx-Driver-Samples#11]
i can't find any specification for RTL8168D's registers. Just found for RTL8168B, they are similar but this device doesn't allow receive side scaling. There's an online spec or another compatible dev to follow yours driver sample. thanks, great work.

**Comment by woreno** [comment]
Well the registers after a while are quite understable from the code function; 
The differences from 8168B (the only reference that i could found), are:
- 3 more IMR/ISRx registers added for int status and to mask each rx cpu/queue ints
- the "extra 3 RDSAR", the logical base addresses for rx descriptors for each rx cpu/queu can be accessed by the extra pair ERIData/Access registers
- and theres the rssIndirectionTable vector needed to perform the cpu handling balancing

But still i'm having doubts about the mapping of the system indirection table found at "EvtAdapterReceiveScalingSetIndirectionEntries@adapter.cpp", appears to be specific for this card. Can anyone point me to a -> reference? thanks!

**Comment by tylerretzlaff** [comment]
Unfortunately you'll have to ask Realtek directly for access to the datasheet it is their intellectual property and I don't have permission to distribute it.

It's been a while since I wrote this code but my recollection is that the datasheet doesn't document the additional IMR/ISRx or RDSAR registers / offsets but they are easy to lookup in the code as you have done.  The datasheet I had only actually described 2 queues (not 4) as the RTL8168D has.

I can comment on the indirection table though since the datasheet I had didn't document it at all (other than the table offset and size).  What values to program it with were initially a mystery but after some thought I basically made an educated guess at how the entries in the table were encoded with knowledge of how many queues the hardware supported.

As for your comment about this indirection table mapping being specific for this card.  I would guess that the mapping/encoding is consistent for the RTL8168 series but the number of entries probably changes depending on how many queues the specific revision has.  To my knowledge Realtek's older NDIS6 driver mostly uses the code for all RTL8168 series adapters with some minor differences depending on specific revisions.

Thanks for checking it out.
